### PR TITLE
[DNM] Docking Umbilicals

### PR DESCRIPTION
**Pull request by bloxgate**
I'm not going to be able to test these until I fix those 510 compile bugs on my end. Anyway, this adds some basic docking umbilicals. Due to some IRL stuff they're going to take a bit longer than I expected, but they will be done before the deadline.
- [x] Transport players
- [x] Air-bridge style thing
- [x] Control console

### CODE REVIEW

**Comment by BDpuffy420** [comment]
I'd recommend you just downgrade to 509.

**Comment by bloxgate** [comment]
@caelaislinn Do you mind taking a quick look at the code, specifically stuff related to actually connecting the two umbilicals. I have a suspicion I may have missed something.

I've just committed a more fleshed out connection code, and have started mapping the actually connection areas that will be created. Mappers/ship design people: I'm thinking two total, one on each side of the ship. Port one would be at the front, starboard to the aft.

File: code/modules/overmap/ships/_ship.dm [hunk 1]
```diff
@@ -199,6 +201,8 @@
 
 	//use pixels per microsecond instead
 	pixel_transform.accelerate_forward(get_acceleration(NORTH))
+	for(var/obj/umbilical_logic/l in umbilicals)
```
Comment (caelaislinn) [inline code/modules/overmap/ships/_ship.dm]: Have a separate list here for connected umbilicals only so it's not looping over all umbilicals every update

File: code/modules/overmap/docking/umbilicals_machinery.dm [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	var/obj/umbilical_logic/connected_to
+	var/deployed = FALSE
+	var/broke = FALSE
+
+/obj/umbilical_logic/New()
+	ship = map_sectors["[z]"] //Need someway to figure out if this is already on a ship
+	ship.umbilicals += src
+	connector.controller = src
+
+/obj/umbilical_logic/proc/attempt_connect(var/obj/umbilical_logic/target, var/other=TRUE)
```
Comment (caelaislinn) [inline code/modules/overmap/docking/umbilicals_machinery.dm]: what calls this? how do players actually trigger the connection and disconnection?

**Comment by caelaislinn** [comment]
Gameplay requirements: 
- Automatically locate nearby viable stations/ships to target (allow player choice if there are multiple)
- Only connects to stuff on the same side as the umbilical (determined by directional facing of its airlock) 
- The receiver cannot override the sender without flying off and snapping the umbilical so it's an alternate method of boarding. 

Good start, missing some stuff.

**Comment by bloxgate** [comment]
Some more testing needs to be done. I'm not going to have a chance to tonight, so if someone else wants to please go ahead and do so.

**Comment by BDpuffy420** [comment]
Has anyone done a test yet?

**Comment by bloxgate** [comment]
Not sure. I can test myself sometime tonight

**Comment by bloxgate** [comment]
There appears to be a problem with the docking computer. Its UI does not open. If someone familiar with Nano UI would mind taking a look at it, it would be appreciated.

**Comment by caelaislinn** [comment]
NanoUI is a hassle to dev. Unlike web apps I've had to relaunch the dmb each time I want to test something iteratively. Apparently there might be some witchery where you can change the template file and have it reload during runtime but I haven't found it yet.

When I get some time I'll dig through.

**Comment by bloxgate** [comment]
Alright thanks

File: nano/templates/docking.tmpl [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<H1>Origin Umbilical</H1>
+{{for data.origins}}
+	<div class="item">
+	{{if data.origin.name == :value.name}} <b>	{{/if}}{{:helper.link(:value.name, null, {'choice': 'origin', 'umbilical' : value.ref}, null}}{{if data.origin.name == :value.name}} </b>	{{/if}}
```
Comment (caelaislinn) [inline nano/templates/docking.tmpl]: indent this properly, oh god. looks like just a problem with your line endings tho. what text editor were you using?

File: nano/templates/docking.tmpl [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+	</div>
+{{empty}}
+	No docking controllers present
+{{/for}}
+
+<H1>Destination Umbilical</H1>
+{{for data.ports}}
+	<div class="item">
+	{{if data.destination.name == :value.name}} <b>	{{/if}}{{:helper.link(:value.name, null, {'choice':'dest', 'umbilical' : vaule.ref}, null}}{{if data.destiantion.name == :value.name}} </b>	{{/if}}
```
Comment (caelaislinn) [inline nano/templates/docking.tmpl]: indents

Comment (caelaislinn) [inline nano/templates/docking.tmpl]: you mispelt "destination" and "value" here. with proper indentation, this was pretty easy to spot :P (both are critical errors that would stop nanoui from opening)

File: nano/templates/docking.tmpl (hunk 3 above)
Comment (bloxgate) [inline nano/templates/docking.tmpl]: Line endings should be normal Unix ones. I used NPP to write it.

File: maps/unsc_cruiser6.dmm [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```
Comment (caelaislinn) [inline maps/unsc_cruiser6.dmm]: whats this new mapfile for?

**Comment by bloxgate** [comment]
Well this is embarrassing, apparently I'd forgot to include the code that opened the GUI in the first place. I'll be committing a fix shortly.

**Comment by caelaislinn** [comment]
Ok I'm going to avoid doing exhaustive testing of this apart from does it work/not work so I'll get you to  write up a guide for future mappers on how to set it up and also just put pick somewhere you think appropriate then put them into the map yourself. I'd suggest fore or midships on deck 4 as there seems to be plenty of room there. 

Also as I mentioned on skype, there's no need to mess around with areas or area procs for moving turfs around. You also don't need somewhere to hide the turfs for the connectors because you can just stuff them inside the entry port when the umbilical is retracted. 

I'd also like it if all the components were moved into the docking folder so they're easier to find.

**Comment by bloxgate** [comment]
As we talked about on Skype, I'm using areas so that everything is the same way it was when its extended. I'd prefer not to have to make code to put every turf into the right spot based on where it was before it was retracted.

I'll move things into the docking folder next chance I get.

**Comment by caelaislinn** [comment]
The code to geometrically lay out a pattern of turfs could be done in a single for or while loop. Reusing functionality for areas because they do a similar thing is not ideal, as it relies on areas. Unless you're gong to be dynamically generating and assigning areas over the source and destination points (at which case you may as well write your own turf move proc) it's going to rely on hardcoded areas and map positions which is not maintainable.

**Comment by bloxgate** [comment]
This is perfectly maintainable. The areas are variables in the controller, and there's no reason they can't be moved around to other places.

**Comment by bloxgate** [comment]
I'm giving up on the current method of doing it, and it looks like my idea for another method doesn't work out too well. Going to redo a large chunk of this soonish.

**Comment by BDpuffy420** [comment]
Update on this?

**Comment by bloxgate** [comment]
Yes, I am still working on this when I get the chance. I have quite a bit of the underlying stuff done, just need to get it to actually work from in game.

**Comment by bloxgate** [comment]
Update: These are almost done. Working out some final bugs, though I'm not having much free time as of late.

**Comment by BDpuffy420** [comment]
Any new updates?

**Comment by bloxgate** [comment]
I have code that should be good for actual docking. Testing was taking a while. Now that startup is reasonable again, I don't think it'll be too much longer once I'm done with school.

**Comment by BDpuffy420** [comment]
Soon!

**Pull request closed without merging**